src/feature/hs/hs_cell.c

```diff
@@ -853,17 +853,46 @@ hs_cell_parse_introduce2(hs_cell_introduce2_data_t *data,
 
   /* First bytes of the ENCRYPTED section are the client public key (they are
    * guaranteed to exist because of the length check above). We are gonna use
-   * the client public key to compute the ntor keys and decrypt the payload: */
-  memcpy(&data->client_pk.public_key, encrypted_section, CURVE25519_PUBKEY_LEN);
-
-  /* Get the right INTRODUCE2 ntor keys and verify the cell MAC */
-  intro_keys = get_introduce2_keys_and_verify_mac(data, encrypted_section,
-                                                  encrypted_section_len);
-  if (!intro_keys) {
-    log_info(LD_REND, "Could not get valid INTRO2 keys on circuit %u "
-             "for service %s", TO_CIRCUIT(circ)->n_circ_id,
-             safe_str_client(service->onion_address));
-    goto done;
+   * the client public key to compute the ntor keys and decrypt the payload:
+   */
+  memcpy(&data->client_pk.public_key, encrypted_section,
+         CURVE25519_PUBKEY_LEN);
+
+  /* If we are configured as an Onion Balance instance, we need to try
+   * validation with the configured master public keys given in the config
+   * file. This is because the master identity key and the blinded key is put
+   * in the INTRODUCE2 cell by the client thus it will never validate with
+   * this instance public key. */
+  if (service->config.ob_master_pubkeys) {
+    hs_cell_introduce2_data_t new_data;
+
+    SMARTLIST_FOREACH_BEGIN(service->config.ob_master_pubkeys,
+                            const ed25519_public_key_t *, pkey) {
+      ed25519_public_key_t blinded_pubkey;
+      uint8_t subcredential[DIGEST256_LEN];
+
+      hs_build_blinded_pubkey(pkey, NULL, 0, hs_get_time_period_num(0),
+                              &blinded_pubkey);
+      hs_get_subcredential(pkey, &blinded_pubkey, subcredential);
+      memcpy(&new_data, data, sizeof(new_data));
+      new_data.subcredential = subcredential;
+      intro_keys = get_introduce2_keys_and_verify_mac(&new_data,
+                                                      encrypted_section,
+                                                      encrypted_section_len);
+      if (intro_keys) {
+        break;
+      }
+    } SMARTLIST_FOREACH_END(pkey);
+  } else {
+    /* Get the right INTRODUCE2 ntor keys and verify the cell MAC */
+    intro_keys = get_introduce2_keys_and_verify_mac(data, encrypted_section,
+                                                    encrypted_section_len);
+    if (!intro_keys) {
```

Comment:
Perhaps this (!intro_keys) case can be moved out of the if clause so that it warns in case the MAC didnt line up in the OB case either.